promql/prettier/prettier.go

```diff
@@ -0,0 +1,577 @@
+// Copyright 2015 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prettier
+
+import (
+	"fmt"
+	"gopkg.in/yaml.v3"
+	"io/ioutil"
+	"reflect"
+	"regexp"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/pkg/rulefmt"
+	"github.com/prometheus/prometheus/promql/parser"
+)
+
+const (
+	// PrettifyRules for prettifying rules files along with the
+	// expressions in them.
+	PrettifyRules = iota
+	// PrettifyExpression for prettifying instantaneous expressions.
+	PrettifyExpression
+	// itemNotFound is used to initialize the index of an item.
+	itemNotFound = -1
+)
+
+// Prettier handles the prettifying and formatting operation over a
+// list of rules files or a single expression.
+type Prettier struct {
+	files      []string
+	expression string
+	Type       uint
+	Node       parser.Expr
+	pd         *padder
+}
+
+type padder struct {
+	indent      string
+	columnLimit int
+	// isNewLineApplied represents a new line was added previously, after
+	// appending the item to the result. It should be set to default
+	// after applying baseIndent. This property is dependent across the padder states
+	// and hence needs to be carried along while formatting.
+	isNewLineApplied bool
+	// isPreviousItemComment confirms whether the previous formatted item was
+	// a comment. This property is dependent on previous padder states and hence needs
+	// to be carried along while formatting.
+	isPreviousItemComment bool
+	// expectAggregationLabels expects the following item(s) to be labels of aggregation
+	// expression. Labels are used in AggregateExpr and in VectorSelector but only the ones
+	// in VectorSelector are splittable.
+	expectAggregationLabels bool
+	// labelsSplittable determines whether the current formatting region lies in labels
+	// matchers in VectorSelector. Since labels in VectorSelectors are only splittable,
+	// is property is applied only after encountering a Left_Brace. This is dependent across
+	// multiple padder states and hence needs to be carried along.
+	labelsSplittable bool
+}
+
+// New returns a new prettier over the given slice of files.
```

Comment:
Promql should not know about files, only expressions.

Reply:
This is intended to be served via Promtool. Hence, the promtool is the calling function mostly, however we can have an API endpoint as well. This is, in fact, the reason why it is kept as a separate module.

Reply:
Then promtool is what should be handling any file handling, not PromQL.

Reply:
Made prettier independent of files. Now, it's handled by promtool.